data/transactions/logic/eval.go

```diff
@@ -612,26 +619,243 @@ func (cx *EvalContext) RunMode() RunMode {
+	// avmBytes in an OpSpec shows that the op pops or yields a []byte
+	avmBytes
 )
 
+func (at avmType) String() string {
+	switch at {
+	case avmNone:
+		return "none"
+	case avmAny:
+		return "any"
+	case avmUint64:
+		return "uint64"
+	case avmBytes:
+		return "[]byte"
+	}
+	return "internal error, unknown type"
+}
+
+var (
+	// StackUint64 is any valid uint64
+	StackUint64 = NewStackType(avmUint64, bound(0, math.MaxUint64))
+	// StackBytes is any valid bytestring
+	StackBytes = NewStackType(avmBytes, bound(0, maxStringSize))
+	// StackAny could be Bytes or Uint64
+	StackAny = StackType{
+		Name:        avmAny.String(),
+		AVMType:     avmAny,
+		ValueBound:  StackUint64.ValueBound,
+		LengthBound: StackBytes.LengthBound,
+	}
+	// StackNone is used when there is no input or output to
+	// an opcode
+	StackNone = StackType{
+		Name:    avmNone.String(),
+		AVMType: avmNone,
+	}
+
+	// StackBoolean constrains the int to 1 or 0, representing True or False
+	StackBoolean = NewStackType(avmUint64, bound(0, 1), "bool")
+	// StackAddress represents an address
+	StackAddress = NewStackType(avmBytes, static(32), "addr")
+	// StackBytes32 represents a bytestring that should have exactly 32 bytes
+	StackBytes32 = NewStackType(avmBytes, static(32), "[32]byte")
+	// StackBigInt represents a bytestring that should be treated like an int
+	StackBigInt = NewStackType(avmBytes, bound(0, maxByteMathSize), "bigint")
+	// StackMethodSelector represents a bytestring that should be treated like a method selector
+	StackMethodSelector = NewStackType(avmBytes, static(4), "method")
+	// StackStorageKey represents a bytestring that can be used as a key to some storage (global/local/box)
+	StackStorageKey = NewStackType(avmBytes, bound(0, 64), "key")
+	// StackBoxKey represents a bytestring that can be used as a key to a box
+	StackBoxKey = NewStackType(avmBytes, bound(1, 64), "bkey")
+
+	// AllStackTypes is a list of all the stack types we recognize
+	// so that we can iterate over them in doc prep
+	AllStackTypes = []StackType{
+		StackUint64,
+		StackBytes,
+		StackAny,
+		StackNone,
+		StackBoolean,
+		StackBytes32,
+		StackBigInt,
+		StackMethodSelector,
+		StackStorageKey,
+		StackBoxKey,
+	}
+)
+
+func bound(min, max uint64) [2]uint64 {
+	return [2]uint64{min, max}
+}
+
+func static(size uint64) [2]uint64 {
+	return bound(size, size)
+}
+
+// StackType describes the type of a value on the operand stack
+type StackType struct {
+	Name        string
+	AVMType     avmType
+	LengthBound [2]uint64
+	ValueBound  [2]uint64
+}
+
+// NewStackType Initializes a new StackType with fields passed
+func NewStackType(at avmType, bounds [2]uint64, stname ...string) StackType {
+	name := at.String()
+	if len(stname) > 0 {
+		name = stname[0]
+	}
+
+	st := StackType{Name: name, AVMType: at}
+
+	switch at {
+	case avmBytes:
+		st.LengthBound = bounds
+	case avmUint64:
+		st.ValueBound = bounds
+	}
+
+	return st
+}
+
+func unionStackTypes(a, b StackType) (StackType, error) {
```

Comment:
Maybe this should be a method on StackType?
```suggestion
func (a StackType) union(b StackType) (StackType, error) {
```